File: src/jp/lionas/alexa/fundra/AnswerNoIntentHandler.java
package jp.lionas.alexa.fundra;

import com.amazon.ask.dispatcher.request.handler.HandlerInput;
import com.amazon.ask.dispatcher.request.handler.RequestHandler;
import com.amazon.ask.model.Response;

import java.util.Map;
import java.util.Optional;

import static com.amazon.ask.request.Predicates.intentName;
import static com.amazon.ask.request.Predicates.sessionAttribute;

public class AnswerNoIntentHandler implements RequestHandler {

    @Override
    public boolean canHandle(HandlerInput handlerInput) {
        return handlerInput.matches(intentName(Const.ASK_NO_INTENT_NAME)
                .and(sessionAttribute(Const.STATE_KEY, Const.STATE_END).negate()));
    }

    @Override
    public Optional<Response> handle(HandlerInput handlerInput) {

        Map<String, Object> sessionAttributes = handlerInput.getAttributesManager().getSessionAttributes();
        sessionAttributes.put(Const.STATE_KEY, Const.STATE_END);

        return handlerInput.getResponseBuilder()
                .withSpeech(Const.CANCEL_SPEECH_TEXT)
                .withSimpleCard(Const.SKILL_NAME, Const.CANCEL_SPEECH_TEXT)
                .withShouldEndSession(true)
                .build();

    }

}

File: src/jp/lionas/alexa/fundra/AnswerNoIntentHandler.kt
package jp.lionas.alexa.fundra

import com.amazon.ask.dispatcher.request.handler.HandlerInput
import com.amazon.ask.dispatcher.request.handler.RequestHandler
import com.amazon.ask.model.Response
import java.util.Optional

import com.amazon.ask.request.Predicates.intentName
import com.amazon.ask.request.Predicates.sessionAttribute
import jp.lionas.alexa.fundra.def.Message
import jp.lionas.alexa.fundra.def.State

class AnswerNoIntentHandler : RequestHandler {

    override fun canHandle(handlerInput: HandlerInput): Boolean {
        return handlerInput.matches(intentName("AMAZON.NoIntent")
                .and(sessionAttribute(State.STATE_KEY, State.STATE_END).negate()))
    }

    override fun handle(handlerInput: HandlerInput): Optional<Response> {

        val sessionAttributes = handlerInput.attributesManager.sessionAttributes
        sessionAttributes[State.STATE_KEY] = State.STATE_END

        return handlerInput.responseBuilder
                .withSpeech(Message.CANCEL_SPEECH_TEXT)
                .withSimpleCard(Message.SKILL_NAME, Message.CANCEL_SPEECH_TEXT)
                .withShouldEndSession(true)
                .build()

    }

}
